thaumato-anakalyptor/ThaumatoAnakalyptor/instances_to_graph.py

```diff
@@ -32,6 +33,72 @@
 sys.path.append('ThaumatoAnakalyptor/sheet_generation/build')
 import sheet_generation
 
+@njit
+def numba_mean_axis0(X):
+    """
+    Compute the mean of a 2D array along axis 0 manually for Numba compatibility.
+    """
+    n_rows, n_cols = X.shape
+    mean_values = np.zeros(n_cols, dtype=X.dtype)
+    
+    for col in range(n_cols):
+        col_sum = 0.0
+        for row in range(n_rows):
+            col_sum += X[row, col]
+        mean_values[col] = col_sum / n_rows
+    
+    return mean_values
+
+@njit
+def numba_norm(arr, axis=1):
+    """
+    Custom implementation of np.linalg.norm for Numba, supporting axis argument.
+    Computes the Euclidean norm along the specified axis.
+    """
+    if axis == 1:
+        # Calculate the norm along rows (for 2D array)
+        return np.sqrt(np.sum(arr ** 2, axis=1))
+    elif axis == 0:
+        # Calculate the norm along columns (for 2D array)
+        return np.sqrt(np.sum(arr ** 2, axis=0))
+    else:
+        raise ValueError("Unsupported axis. Use 0 or 1.")
+    
+@njit
+def geometric_median_(X, eps=1e-5):
```

Comment:
could this be more concise + not needing a new library?
np.exp(np.mean(np.log(arr), axis=axis))